### DISCUSSION TITLE
Support name binding

### DISCUSSION THREAD

**Discussion by oovm** [pest-parser/pest#550] [Ideas]
## Motivation

Name-binding refers to the name of a node in the specified syntax tree, for example `=` in [ANTLR](https://github.com/antlr/antlr4/blob/master/doc/parser-rules.md#rule-element-labels), `field` in [tree-sitter](https://tree-sitter.github.io/tree-sitter/creating-parsers#using-fields).

Very useful when the name is repeated, such as distinguishing the left and right nodes of an expression.

```rs
add = {lhs = expr ~ "+" ~ rhs = epxr}
```

## Design

https://github.com/pest-parser/pest/blob/f0a85d1daed8a45e6be136b81df06f2a43299ef8/pest/src/iterators/pairs.rs#L34-L39

```rs
pub struct Pairs<'i, R> {
    // ...others    
    mark: Option<&'i str>
}
```

Then you can use a very convenient interface to find the node you want.

```rs
pairs.find_first_mark(mark: &str) -> Option<&Self>
pairs.find_marks(mark: &str)-> Iter<&Self>
```

The current process is more cumbersome.

```rs
let mut count_expr = 0
for pair in paris { 
    if else if else if else... 
    count_expr += 1
}
```

---

## Specific changes

https://github.com/pest-parser/pest/blob/f0a85d1daed8a45e6be136b81df06f2a43299ef8/pest/src/parser_state.rs#L77-L86

Here to extract the value from the state:

```rs
let len = state.queue.len(); 
let mark= std::mem::take(state.mark); 
Ok(pairs::new(Rc::new(state.queue), input, 0, len, mark)) 
```

And add a primitive operation here:

```rs
impl<'i, R: RuleType> ParserState<'i, R> {
    pub fn mark(mut self: Box<Self>, s: &str) -> ParseResult<Box<Self>> {
        self.mark = Some(s);
        Ok(Self)
    }
}
```

Ideally it should be able to correspond to such code:

```rs
add = {lhs = expr ~ "+" ~ rhs = epxr}
```

```rs
pub fn add(state: Box<pest::ParserState<Rule>>) -> pest::ParseResult<Box<pest::ParserState<Rule>>> {
    state.rule(Rule::add, |state| {
        state.sequence(|state| {
            state
                .and_then(|state| self::expr(state))
                .and_then(|state| state.mark("lhs"))
                .and_then(|state| super::hidden::skip(state))
                .and_then(|state| state.match_string("+"))
                .and_then(|state| super::hidden::skip(state))
                .and_then(|state| self::expr(state))
                .and_then(|state| state.mark("rhs"))
        })
    })
}
```

## Compatibility

Does not break existing features.

**Comment by NoahTheDuke** [comment]
This is a very cool idea. Python's new PEG parser has the same feature, detailed [here](https://www.python.org/dev/peps/pep-0617/#variables-in-the-grammar).

**Comment by oovm** [comment]
Similarly, we can also add the branch tag

```rs
expr = {
    "(" ~ expr ~ ")"                 // branch: None
  | term ~ ("/"|"*") ~ term  #Mul    // branch: Some("Mul")
  | term ~ ("+"|"-") ~ term  #Add    // branch: Some("Add")
}
```

which call after `or_else`

**Reply by NoahTheDuke** [reply]
How would this be used? Any examples?

**Reply by oovm** [reply]
Basically, I've add the doc test for `tag_branch`, and here's a real world usage see: [ygg-bootstrap/parse.rs](https://github.com/ygg-lang/yggdrasil-rs/blob/82cfeb8db1c96d42d4d006e7d19ca010f77942c8/projects/ygg-bootstrap/src/cst/parse.rs#L187-L198)

**Comment by CAD97** [comment]
This seems reasonable to me! It's a nice middle-ground between pest's fully "untyped" parse tree and something more strongly typed like the [pest-ast draft](https://github.com/pest-parser/ast) or the mythical pest@3.

**Comment by tomtau** [comment]
I was looking into reviving https://github.com/pest-parser/pest/pull/552 and also looked into extending the meta grammar.
The original PR #552 is fine, but extending the meta grammar is a bit tricky: 1) it seems it can't be done in a way that does not break a semver (but that should be ok, as those meta enums are now documented with a warning that this may happen, given it happened in the past), 2) interactions with the optimizer may be strange... for example, the optimizer will rewrite `"a" ~ "b"` to `"ab"`, but what if there're node tags on those terminals?

@oovm @CAD97 @NoahTheDuke have you looked into or thought about the meta grammar implementation? Given how the code is written, there are a few ways:

- adding a tag field to each `Expr` (and its follow-up near duplicate enums) variants (yuck?)
- rewriting those generator methods that operate on `Expr` (or other related enums) to use some wrapper struct (`TaggedExpr`) instead
- a bigger logical refactoring, e.g. to use something like https://github.com/guygastineau/rust-trees-that-grow
- adding a new `Tag` variant to `Expr`, propagating it to related enums and patching the generator to look for that

I tried the first one but it's likely a no-go... I'm thinking the last may be the path of least resistance, as the explicit variant may fix up the potential issues with the optimizer... anyway, any thoughts? or should this be left for a simpler v3 grammar/implementation?

**Reply by NoahTheDuke** [reply]
This change (however it's accomplished) is big enough that we should probably do a semver bump, so that feels okay to me. I left some comments in #832 about your implementation.
